Firestore/core/src/firebase/firestore/api/listener_registration.mm

```diff
@@ -17,21 +17,24 @@
 #include "Firestore/core/src/firebase/firestore/api/listener_registration.h"
 
 #import "Firestore/Source/Core/FSTFirestoreClient.h"
-#import "Firestore/Source/Util/FSTAsyncQueryListener.h"
 
 namespace firebase {
 namespace firestore {
 namespace api {
 
 void ListenerRegistration::Remove() {
-  [async_listener_ mute];
-  async_listener_ = nil;
+  auto async_listener = async_listener_.lock();
+  if (async_listener) {
+    async_listener->Mute();
+    async_listener_.reset();
+    async_listener.reset();
```

Comment:
Reseting this local variable has the effect of making the QueryListener the only thing keeping the AsyncEventListener alive, so that it can be destroyed inline with QueryListener.

However, now that I think about it, there's no great way to guarantee that the QueryListener will be destroyed on our queue because the query_listener local retains it. This means that the various EventListener destructors will run on this thread.

In the end I think that's not as bad a thing as I had thought it would be. The QueryListener and the AsyncEventListener are created on the user's thread so it's fine for them to be destroyed there.

Secondarily, now that you've prompted about it, I think that trying to manage which thread destroys an object by carefully managing the lifetime of shared_ptrs is a losing game. If we really care we need to make this contractual, i.e. that the pointed to object, in its destructor submits the real destructor to an executor. That's beyond the scope of this change though, so I'm not doing anything about it here.

So, I've removed the extra resets.